tests/h_item/h_item_api.cpp

```diff
@@ -19,26 +23,78 @@ struct h_item_api_kernel_common;
  * @tparam currentDim Current dimension to get the value for
  */
 template <int numDims, int currentDim>
-struct h_item_api_kernel_single {
+struct kernel_single {
+  using success_acc_t =
+      cl::sycl::accessor<bool, 1, cl::sycl::access::mode::write>;
+
   cl::sycl::range<numDims> kernelLogicalLocalRange;
+  success_acc_t success_acc;
 
   void operator()(cl::sycl::group<numDims> group) const {
     group.parallel_for_work_item(
         kernelLogicalLocalRange, [&](cl::sycl::h_item<numDims> item) {
-          size_t globalRange = item.get_global_range(currentDim);
-          size_t globalId = item.get_global_id(currentDim);
-          size_t localRange = item.get_local_range(currentDim);
-          size_t localId = item.get_local_id(currentDim);
-          size_t logicalLocalRange = item.get_logical_local_range(currentDim);
-          size_t logicalLocalId = item.get_logical_local_id(currentDim);
-          size_t physicalLocalRange = item.get_physical_local_range(currentDim);
-          size_t physicalLocalId = item.get_physical_local_id(currentDim);
+          bool success = true;
+          {
+            auto value = item.get_global_range(currentDim);
+            auto expected = item.get_global_range()[currentDim];
+            success &= value == expected;
+          }
+          {
+            auto value = item.get_global_id(currentDim);
+            auto expected = item.get_global_id()[currentDim];
+            success &= value == expected;
+          }
+          {
+            auto value = item.get_local_range(currentDim);
+            auto expected = item.get_local_range()[currentDim];
+            success &= value == expected;
+          }
+          {
+            auto value = item.get_local_id(currentDim);
+            auto expected = item.get_local_id()[currentDim];
+            success &= value == expected;
+          }
+          {
+            auto value = item.get_logical_local_range(currentDim);
+            auto expected = item.get_logical_local_range()[currentDim];
+            success &= value == expected;
+          }
+          {
+            auto value = item.get_logical_local_id(currentDim);
+            auto expected = item.get_logical_local_id()[currentDim];
+            success &= value == expected;
+          }
+          {
+            auto value = item.get_physical_local_range(currentDim);
+            auto expected = item.get_physical_local_range()[currentDim];
+            success &= value == expected;
+          }
+          {
+            auto value = item.get_physical_local_id(currentDim);
+            auto expected = item.get_physical_local_id()[currentDim];
+            success &= value == expected;
+          }
+
+          if (!success) success_acc[0] = false;
```

Comment:
```suggestion
         *success_acc = success;
```

Reply:
We should change value from `true` to `false` only. It's required to avoid data race. I've added comment to make it explicit in e16addc. Please, review.

Reply:
Oh, I see. Actually it is not even about a race condition but just about the algorithm that I completely got wrong... :-(
But since you are mentioning race conditions, I have the feeling that in OpenCL it is UB to have different work-item writing at the same place, even if they are writing the same value. I guess too that it will work anyway.
Good news, in SYCL 2020 we have reductions for this... :-)

Reply:
I guess this comment goes to the IDs storage logic also. Just to make sure - should I change algorithm?

Reply:
Yes, that is a problem...
Let's discuss this during next SYCL committee meeting and look at where the discussion goes before changing anything.

Reply:
I think @keryell is right that this is UB. This is a race condition in C++ memory model.
@yuriykoch, let's allocate one result element per WI and reduce on the host.

Reply:
I agree with @bader and @keryell.

It's a little surprising, but this is UB according to http://eel.is/c++draft/intro.multithread#intro.races-21:

> The execution of a program contains a data race if it contains two potentially concurrent conflicting actions, at least one of which is not atomic, and neither happens before the other, except for the special case for signal handlers described below. Any such data race results in undefined behavior.

This [paper](https://www.usenix.org/legacy/events/hotpar11/tech/final_files/Boehm.pdf) by Hans Boehm also calls out the issue being discussed here explicitly as being problematic:

> ...all kinds of C or C++ source-level "benign" races discussed in the literature can in fact lead to incorrect execution as a result of perfectly reasonable compiler transformations, or when the program is moved to a different hardware platform. ... Perhaps most surprisingly, this includes even the case of potentially concurrent writes of the same value by different threads.

Reply:
Yes, this is an interesting (scary) paper from Hans Boehm. For more citation context it comes from https://www.usenix.org/conference/hotpar-11/how-miscompile-programs-benign-data-races

Reply:
Just to make sure - couldn't we just use atomic access mode to ensure there would be no UB nevertheless if the tests succeeds or fails due - for example - of some duplicated global or logical IDs?
I mean:
  - we do also have the same problem for global_acc, logical_acc and physical_acc
  - for a case of invalid global linear id there is no way to tell how to avoid UB due to writing at the same offset from within WI
  - strictly speaking UB is not limited to the broken data values or abort() call or anything, so we cannot expect test to have any valid output in case of duplicated IDs
  - there were a plenty of tests using something like "if (!success) accessor[0] = false" originally, so may be we should just use atomic when we see it to quick-fix without test logic changes

What do you think?

Reply:
We could but atomics are an optional feature...

Reply:
Is it optional for the int/uint though? We could fit into these types instead of bool and size_t with some additional checks on ID values due possible truncation of invalid ones.

Reply:
I don't think it's a good idea to use features that are not supposed to be covered by the test and if they can be avoided.
Adding dependency on supporting atomics here seems unnecessary to me.
@yuriykoch, what problems do you see with the approach proposed above?

Reply:
@bader I agree, that's a huge cons against atomic usage.
The main issue is that we don't remove UB completely. We are testing h_item API, so we would expect test to provide some output in case there are invalid IDs. For example, in case all work items return zero ID there still will be UB by writing at the same zero location. So we may not expect any proper failure report in such case.
If that's not critical I would use logical local ID and group ID to provide unique offset for every execution to avoid failure silencing at least.

Reply:
The main point of this discussion is to remove UB from **the test code**. If UB exists in `h_item` **implementation**, there is nothing we can do about it and using `atomics` doesn't help neither.

Reply:
Done by 555af6886abed93af319a979a794e42db370268f. I've tried my best to describe all details and caveats of test implementation within the commit message